runtime/tcp.cpp

```diff
@@ -0,0 +1,346 @@
+    opened_tcp_client_sockets->set_value(stream, fdopen(fd, "r"));
+  }
+  FILE *src = opened_tcp_client_sockets->get_value(stream);
+  int res = fgetc(src);
+  dl::leave_critical_section();
+
+  if (res == -1) {
+    return false;
+  }
+  return f$chr(res);
+}
+
+bool tcp_fclose(const Stream &stream) {
+  int fd = tcp_get_fd(stream);
+
+  dl::enter_critical_section();
+  int res = close(fd);
+  if (opened_tcp_client_sockets->get_value(stream)) {
+    fclose(opened_tcp_client_sockets->get_value(stream));
+  }
+  opened_tcp_client_sockets->unset(stream);
+  dl::leave_critical_section();
+
+  if (res == -1) {
+    php_warning("fclose() : An error occurred while closing TCP-package");
+    return false;
+  }
+  return true;
+}
+
+bool tcp_feof(const Stream &stream) {
+  int fd = tcp_get_fd(stream);
+  if (fd == -1) {
+    return false;
+  }
+
+  dl::enter_critical_section();
+  if (!opened_tcp_client_sockets->get_value(stream)) {
+    opened_tcp_client_sockets->set_value(stream, fdopen(fd, "r"));
+  }
+  FILE *src = opened_tcp_client_sockets->get_value(stream);
+  int res = feof(src);
+  dl::leave_critical_section();
+
+  return res != 0;
+}
+
+/*
+ * now available only blocking/non-blocking option
+ * */
+bool tcp_stream_set_option(const Stream &stream, int64_t option, int64_t value) {
+  int fd = tcp_get_fd(stream);
+  if (fd == -1) {
+    return false;
+  }
+
+  int flags = fcntl(fd, F_GETFL);
+  switch (option) {
+    case STREAM_SET_BLOCKING_OPTION:
+      int res;
+      if (value) {
+        res = fcntl(fd, F_SETFL, flags & ~O_NONBLOCK);
+      } else {
+        res = fcntl(fd, F_SETFL, flags | O_NONBLOCK);
+      }
+      return res == 0;
+    default:
+      return false;
+  }
+}
+} // namespace
+
+void global_init_tcp_lib() {
+  static stream_functions tcp_stream_functions;
+
+  tcp_stream_functions.name = string("tcp", 3);
+  tcp_stream_functions.fopen = nullptr;
+  tcp_stream_functions.fwrite = tcp_fwrite;
+  tcp_stream_functions.fseek = nullptr;
+  tcp_stream_functions.ftell = nullptr;
+  tcp_stream_functions.fread = tcp_fread;
+  tcp_stream_functions.fgetc = tcp_fgetc;
+  tcp_stream_functions.fgets = tcp_fgets;
+  tcp_stream_functions.fpassthru = nullptr;
+  tcp_stream_functions.fflush = nullptr;
+  tcp_stream_functions.feof = tcp_feof;
+  tcp_stream_functions.fclose = tcp_fclose;
+
+  tcp_stream_functions.file_get_contents = nullptr;
+  tcp_stream_functions.file_put_contents = nullptr;
+
+  tcp_stream_functions.stream_socket_client = tcp_stream_socket_client;
+  tcp_stream_functions.context_set_option = nullptr;
+  tcp_stream_functions.stream_set_option = tcp_stream_set_option;
+  tcp_stream_functions.get_fd = tcp_get_fd;
+
+  register_stream_functions(&tcp_stream_functions, false);
+}
+
+void free_tcp_lib() {
+  dl::enter_critical_section();
+  if (dl::query_num == opened_tcp_client_sockets_last_query_nam) {
+    array<FILE *> *const_opened_udp_sockets = opened_tcp_client_sockets;
```

Comment:
May be '**const_opened_tcp_sockets**' and not '**const_opened_udp_sockets**'? :)

Reply:
fix